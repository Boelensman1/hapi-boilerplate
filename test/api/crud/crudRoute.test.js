const path = require('path')
const request = require('supertest')
const setUpHandlerTest = require('test/api/setUpHandlerTest')

const seedLocation = path.join(__dirname, 'crudRoute.seed.yaml')

describe('Test the generated crud routes', () => {
  test('Get list', async () => {
    const { listener } = await setUpHandlerTest(seedLocation)

    const {
      body: { result, pagination, statusCode },
    } = await request(listener)
      .get('/crud/role')
      .query({ sort: JSON.stringify({ field: 'id', order: 'desc' }) })
      .query({ filter: JSON.stringify({ id: [11, 13] }) })
      .query({ range: [0, 1] })
      .expect('Content-Type', /json/)
      .expect(200)
      .expect('Content-Range', 'role 0-1/2')

    expect(statusCode).toBe(200)
    expect(result).toHaveLength(1)
    expect(result[0]).toMatchSnapshot({
      createdAt: expect.any(String),
    })
    expect(pagination).toMatchSnapshot()
  })
  test('Get filtered list', async () => {
    const { listener } = await setUpHandlerTest(seedLocation)

    const {
      body: { result, pagination, statusCode },
    } = await request(listener)
      .get('/crud/role')
      .query({ filter: JSON.stringify({ id: [11, 13], name: 'nameB' }) })
      .expect('Content-Type', /json/)
      .expect(200)
      .expect('Content-Range', 'role 0-1/1')

    expect(statusCode).toBe(200)
    expect(result).toHaveLength(1)
    expect(result[0]).toMatchSnapshot({
      createdAt: expect.any(String),
    })
    expect(pagination).toMatchSnapshot()
  })

  test('Get one', async () => {
    const { listener } = await setUpHandlerTest(seedLocation)

    const {
      body: { result, statusCode },
    } = await request(listener)
      .get('/crud/role/12')
      .expect('Content-Type', /json/)
      .expect(200)

    expect(statusCode).toBe(200)
    expect(result).toMatchSnapshot({
      id: 12,
      createdAt: expect.any(String),
    })

    await request(listener)
      .get('/crud/role/999')
      .expect('Content-Type', /json/)
      .expect(404)
  })

  test('Get one with relations', async () => {
    const { listener } = await setUpHandlerTest(seedLocation)

    const {
      body: { result, statusCode },
    } = await request(listener)
      .get('/crud/user/100')
      .query({ getRelations: true })
      .expect('Content-Type', /json/)
      .expect(200)

    expect(statusCode).toBe(200)
    expect(result).toMatchSnapshot({
      id: 100,
      createdAt: expect.any(String),
      role: {
        createdAt: expect.any(String),
      },
    })
  })

  test('Get relation', async () => {
    const { listener } = await setUpHandlerTest(seedLocation)

    const {
      body: { result, statusCode },
    } = await request(listener)
      .get('/crud/role/10/users')
      .expect('Content-Type', /json/)
      .expect(200)

    expect(statusCode).toBe(200)
    expect(result[0]).toMatchSnapshot({
      createdAt: expect.any(String),
    })

    // role does not exists
    await request(listener)
      .get('/crud/role/999/user')
      .expect('Content-Type', /json/)
      .expect(404)

    // has no such relation
    await request(listener)
      .get('/crud/role/0/nonExistingRelation')
      .expect('Content-Type', /json/)
      .expect(404)
  })

  test('Delete one', async () => {
    const { listener, models } = await setUpHandlerTest(seedLocation)

    const {
      body: { statusCode },
    } = await request(listener)
      .delete('/crud/role/12')
      .expect('Content-Type', /json/)
      .expect(200)

    expect(statusCode).toBe(200)
    const role = await models.role.query().findById(12)
    expect(role).toBeUndefined()
  })

  test('Delete non existing should return 404', async () => {
    const { listener } = await setUpHandlerTest(seedLocation)

    await request(listener)
      .delete('/crud/role/999')
      .expect('Content-Type', /json/)
      .expect(404)
  })

  test('Create one', async () => {
    const { listener, models } = await setUpHandlerTest(seedLocation)

    const payload = {
      name: 'New Role',
    }

    const {
      body: { statusCode, result },
    } = await request(listener)
      .post('/crud/role')
      .send(payload)
      .expect('Content-Type', /json/)
      .expect(201)

    expect(statusCode).toBe(201)
    const role = await models.role.query().findById(result.id)
    expect(role).toEqual(expect.objectContaining(payload))
    expect(role).toMatchSnapshot({
      createdAt: expect.any(Date),
    })

    result.createdAt = new Date(result.createdAt)
    expect(result).toEqual(role.toJSON())
  })

  test('Update one', async () => {
    const { listener, models } = await setUpHandlerTest(seedLocation)

    const payload = {
      name: 'New Name',
    }

    const {
      body: { statusCode, result },
    } = await request(listener)
      .put('/crud/role/10')
      .send(payload)
      .expect('Content-Type', /json/)
      .expect(200)

    expect(statusCode).toBe(200)
    const role = await models.role.query().findById(result.id)
    expect(role).toEqual(expect.objectContaining(payload))
    expect(role).toMatchSnapshot({
      id: 10,
      updatedAt: expect.any(Date),
      createdAt: expect.any(Date),
    })

    result.createdAt = new Date(result.createdAt)
    result.updatedAt = new Date(result.updatedAt)
    expect(result).toEqual(role.toJSON())
  })

  test('Update one with relations', async () => {
    const { listener, seed } = await setUpHandlerTest(seedLocation)

    const payload = {
      name: 'New Name',
      users: [
        {
          id: 100,
          username: 'New UserName',
          passwordHash: seed.user[0].passwordHash,
        },
      ],
    }

    const {
      body: { result, statusCode },
    } = await request(listener)
      .put('/crud/role/10')
      .send(payload)
      .expect('Content-Type', /json/)
      .expect(200)

    expect(statusCode).toBe(200)
    expect(result).toMatchSnapshot({
      id: 10,
      name: 'New Name',
      updatedAt: expect.any(String),
      createdAt: expect.any(String),
      users: [
        {
          username: 'New UserName',
          updatedAt: expect.any(String),
          createdAt: expect.any(String),
        },
      ],
    })
  })
})
