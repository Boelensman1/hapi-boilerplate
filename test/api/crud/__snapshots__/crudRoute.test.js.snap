// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the generated crud routes Create one 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 14,
  "name": "New Role",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test the generated crud routes Get filtered list 1`] = `
Object {
  "createdAt": Any<String>,
  "id": 13,
  "name": "nameB",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test the generated crud routes Get filtered list 2`] = `
Object {
  "count": 1,
}
`;

exports[`Test the generated crud routes Get list 1`] = `
Object {
  "createdAt": Any<String>,
  "id": 13,
  "name": "nameB",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test the generated crud routes Get list 2`] = `
Object {
  "count": 2,
  "next": "https://www.example.org/crud/role/?sort=%7B%22field%22%3A%22id%22%2C%22order%22%3A%22desc%22%7D&filter=%7B%22id%22%3A%5B11%2C13%5D%7D&range=%5B1%2C2%5D",
}
`;

exports[`Test the generated crud routes Get one 1`] = `
Object {
  "createdAt": Any<String>,
  "id": 12,
  "name": "nameA",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test the generated crud routes Get one with relations 1`] = `
Object {
  "createdAt": Any<String>,
  "id": 100,
  "loggedInAt": null,
  "role": Object {
    "createdAt": Any<String>,
    "id": 10,
    "name": "nameC",
    "posts_create": 0,
    "posts_delete": 0,
    "posts_read": 0,
    "posts_update": 0,
    "roles_create": 0,
    "roles_delete": 0,
    "roles_read": 0,
    "roles_update": 0,
    "updatedAt": null,
    "users_create": 0,
    "users_delete": 0,
    "users_read": 0,
    "users_update": 0,
  },
  "roleId": 10,
  "session": Array [],
  "updatedAt": null,
  "username": "Relation User",
}
`;

exports[`Test the generated crud routes Get relation 1`] = `
Object {
  "createdAt": Any<String>,
  "id": 100,
  "loggedInAt": null,
  "roleId": 10,
  "updatedAt": null,
  "username": "Relation User",
}
`;

exports[`Test the generated crud routes Update one 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 10,
  "name": "New Name",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": Any<Date>,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test the generated crud routes Update one with relations 1`] = `
Object {
  "createdAt": Any<String>,
  "id": 10,
  "name": "New Name",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": Any<String>,
  "user": Array [
    Object {
      "createdAt": Any<String>,
      "id": 100,
      "loggedInAt": null,
      "roleId": 10,
      "updatedAt": Any<String>,
      "username": "New UserName",
    },
  ],
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;
